### PR TITLE
feat(autocomplete): add contextual skip logic for smart autocomplete triggering

### PR DESCRIPTION
**Pull request by markijbema**
## Summary

Implements intelligent autocomplete skip logic to reduce unnecessary LLM requests and improve user experience by avoiding autocomplete in inappropriate contexts.

## Key Features

- **Language-aware statement terminator detection** - Supports 10+ language families with customizable terminators
- **Skip at end of complete statements** - Avoids autocomplete after `;`, `}`, `)`, etc.
- **Skip during mid-word typing** - Prevents interruptions when typing longer identifiers (>2 chars)
- **Comprehensive test coverage** - 313 test cases covering edge cases and language-specific behavior
- **Centralized configuration** - Easy to add support for new languages

### New Module: `contextualSkip.ts`

Provides a single entry point `shouldSkipAutocomplete()` that checks:
1. Whether cursor is at end of a complete statement
2. Whether user is mid-word typing (and word is long enough to skip)

### Language Support

- **C-like languages**: JavaScript, TypeScript, C/C++, C#, Java, Kotlin, Scala, Swift, PHP, Dart, CSS/SCSS/Less, JSON
- **Python**: Custom terminators for Python syntax
- **Ruby**: Including `end` keyword support
- **Go, Rust**: Standard C-like terminators
- **Shell**: Including `fi`, `done`, `esac` keywords
- **SQL**: Semicolon terminators
- **Lisp family**: Clojure, Scheme, Elisp, Racket
- **Markup**: HTML, XML, SVG, Vue, Svelte

## Examples

**Autocomplete skipped (no LLM request):**
```typescript
console.log('foo');<CURSOR>
const x = 5;<CURSOR>
}<CURSOR>
function test() { return true; }<CURSOR>
```

**Autocomplete still triggers:**
```typescript
console.<CURSOR>log();  // mid-line, content after cursor
const x =<CURSOR>       // incomplete assignment
function test() {<CURSOR>  // opening brace, expecting body
con<CURSOR>             // short word (≤2 chars)
```

## Changes

- Add `contextualSkip.ts` module with `shouldSkipAutocomplete()` entry point
- Integrate skip logic into `GhostInlineCompletionProvider`
- Add extensive test suite in `contextualSkip.spec.ts` (313 tests)
- Language-specific configuration for easy expansion

## Testing

All tests pass:
- 313 test cases in `contextualSkip.spec.ts`
- Covers edge cases, language-specific behavior, and boundary conditions
- Tests for both skip and trigger scenarios

### CODE REVIEW

**Comment by markijbema** [comment]
Added minor changeset so users see stuff is being done, and so that we can release it

File: src/services/ghost/classic-auto-complete/contextualSkip.ts [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+
+	const wordMatch = prefix.match(/([a-zA-Z_][a-zA-Z0-9_]*)$/)
+	const lengthOfWordAtEndOfPrefix = wordMatch ? wordMatch[1].length : 0
+
+	return lengthOfWordAtEndOfPrefix > 2 && !suffixStartsWithWordChar
+}
+
+export function shouldSkipAutocomplete(prefix: string, suffix: string, languageId?: string): boolean {
+	return isAtEndOfStatement(prefix, suffix, languageId) || isMidWordTyping(prefix, suffix)
+}
```
Comment (EamonNerbonne) [inline src/services/ghost/classic-auto-complete/contextualSkip.ts:136]: This file feels too complex.

File: src/services/ghost/classic-auto-complete/contextualSkip.ts [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+export function getTerminatorsForLanguage(languageId: string): Set<string> {
+	const family = LANGUAGE_FAMILY_MAP[languageId]
+	if (family) {
+		const config = LANGUAGE_CONFIGS.find((c) => c.family === family)
+		if (config) {
+			return new Set(config.terminators)
+		}
+	}
+	return DEFAULT_TERMINATORS
+}
```
Comment (EamonNerbonne) [inline src/services/ghost/classic-auto-complete/contextualSkip.ts]: e.g. this includes a fallback...

File: src/services/ghost/classic-auto-complete/contextualSkip.ts [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+	const currentLinePrefix = prefixLines[prefixLines.length - 1]
+	const trimmedLinePrefix = currentLinePrefix.trimEnd()
+
+	// Empty line - not at end of statement
+	if (trimmedLinePrefix.length === 0) {
+		return false
+	}
+
+	// Get language-specific terminators
+	const terminators = languageId ? getTerminatorsForLanguage(languageId) : DEFAULT_TERMINATORS
```
Comment (EamonNerbonne) [inline src/services/ghost/classic-auto-complete/contextualSkip.ts]: but then there's another fallback here...

File: src/services/ghost/classic-auto-complete/contextualSkip.ts [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+
+	// Check single-character terminators
+	const lastChar = trimmedLinePrefix[trimmedLinePrefix.length - 1]
+	if (terminators.has(lastChar)) {
```
Comment (EamonNerbonne) [inline src/services/ghost/classic-auto-complete/contextualSkip.ts]: and we check for single char strings here...

File: src/services/ghost/classic-auto-complete/contextualSkip.ts [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+		languages: ["html", "xml", "svg", "vue", "svelte"],
+		terminators: [],
+	},
+]
+
+/**
+ * Maps VS Code language IDs to language families.
+ * Computed from LANGUAGE_CONFIGS for efficient lookup.
+ */
+const LANGUAGE_FAMILY_MAP: Record<string, string> = LANGUAGE_CONFIGS.reduce(
```
Comment (EamonNerbonne) [inline src/services/ghost/classic-auto-complete/contextualSkip.ts]: this feels oddly indirect - we got from lang->family->terminators, but with an intermediate lookup that's only every built and used from a more direct data structure, the configs.  why not just a direct lang->terminators lookup?

The whole concept "family" seems redundant.

File: src/services/ghost/classic-auto-complete/contextualSkip.ts [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+	const suffixStartsWithWordChar = suffix.length > 0 && /[a-zA-Z0-9_]/.test(suffix[0])
+
+	const wordMatch = prefix.match(/([a-zA-Z_][a-zA-Z0-9_]*)$/)
+	const lengthOfWordAtEndOfPrefix = wordMatch ? wordMatch[1].length : 0
+
+	return lengthOfWordAtEndOfPrefix > 2 && !suffixStartsWithWordChar
+}
+
+export function shouldSkipAutocomplete(prefix: string, suffix: string, languageId?: string): boolean {
+	return isAtEndOfStatement(prefix, suffix, languageId) || isMidWordTyping(prefix, suffix)
```
Comment (EamonNerbonne) [inline src/services/ghost/classic-auto-complete/contextualSkip.ts:135]: we should skip autocomplete after e.g. `)` in most languages?  Really?

Comment (markijbema) [reply src/services/ghost/classic-auto-complete/contextualSkip.ts:135]: i went back and forth on this, but i think it is ok, if it is not the end of the line you will type another character. Both false positives and false negatives aren't that important i think

File: src/services/ghost/classic-auto-complete/contextualSkip.ts (hunk 3 above)
Comment (markijbema) [reply src/services/ghost/classic-auto-complete/contextualSkip.ts]: oh that's idiotic, we can just pass in null/undefined

File: src/services/ghost/classic-auto-complete/contextualSkip.ts (hunk 4 above)
Comment (markijbema) [reply src/services/ghost/classic-auto-complete/contextualSkip.ts]: and strings later; it has slightly better performance, you gather that's not necessary here?

File: src/services/ghost/classic-auto-complete/contextualSkip.ts (hunk 5 above)
Comment (markijbema) [reply src/services/ghost/classic-auto-complete/contextualSkip.ts]: the idea is that we'll have more data per family later, in an intermediate version i had. However we can make that lookup for now, I just like the administration we have

File: src/services/ghost/classic-auto-complete/contextualSkip.ts [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+	// Get the last character before the cursor (excluding trailing whitespace)
+	const prefixLines = prefix.split("\n")
+	const currentLinePrefix = prefixLines[prefixLines.length - 1]
+	const trimmedLinePrefix = currentLinePrefix.trimEnd()
+
+	// Empty line - not at end of statement
+	if (trimmedLinePrefix.length === 0) {
+		return false
+	}
+
```
Comment (EamonNerbonne) [inline src/services/ghost/classic-auto-complete/contextualSkip.ts:107]: effectively dead code.

File: src/services/ghost/classic-auto-complete/contextualSkip.ts [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+}
+
+function isAtEndOfStatement(prefix: string, suffix: string, languageId?: string): boolean {
+	// Get the current line's content after the cursor
+	const suffixFirstLine = suffix.split("\n")[0]
+
+	// If there's non-whitespace content after the cursor, we're not at end of line
+	if (suffixFirstLine.trim().length > 0) {
+		return false
+	}
```
Comment (EamonNerbonne) [inline src/services/ghost/classic-auto-complete/contextualSkip.ts:96]: This might be correct, but it's not intuitively obvious to me why the suffix impacts whether you're at the end of a statement.

File: src/services/ghost/classic-auto-complete/contextualSkip.ts [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+
+	return false
+}
+
+function isMidWordTyping(prefix: string, suffix: string): boolean {
+	if (prefix.length === 0) {
+		return false
+	}
+
+	const suffixStartsWithWordChar = suffix.length > 0 && /[a-zA-Z0-9_]/.test(suffix[0])
```
Comment (EamonNerbonne) [inline src/services/ghost/classic-auto-complete/contextualSkip.ts]:
```suggestion
	const suffixStartsWithWordChar = /^[a-zA-Z0-9_]/.test(suffix)
```

?

File: src/services/ghost/classic-auto-complete/contextualSkip.ts [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+
+	const wordMatch = prefix.match(/([a-zA-Z_][a-zA-Z0-9_]*)$/)
+	const lengthOfWordAtEndOfPrefix = wordMatch ? wordMatch[1].length : 0
+
+	return lengthOfWordAtEndOfPrefix > 2 && !suffixStartsWithWordChar
```
Comment (EamonNerbonne) [inline src/services/ghost/classic-auto-complete/contextualSkip.ts]: shouldn't this be 
```suggestion
	return lengthOfWordAtEndOfPrefix > 2 && suffixStartsWithWordChar
```

?

(boolean logic and all, maybe I'm missing something obvious)

**Review by EamonNerbonne — APPROVED** [review]
Some sanity check ideas that might be off base; use your judgement!

File: src/services/ghost/classic-auto-complete/contextualSkip.ts (hunk 8 above)
Comment (markijbema) [reply src/services/ghost/classic-auto-complete/contextualSkip.ts:96]: if there's code after the cursor we are not at the end of the line (i'm not trying to handle multiple statements on a line / inline editting with this)

File: src/services/ghost/classic-auto-complete/contextualSkip.ts (hunk 10 above)
Comment (markijbema) [reply src/services/ghost/classic-auto-complete/contextualSkip.ts]: no, but it should be different, thx

File: src/services/ghost/classic-auto-complete/contextualSkip.ts (hunk 7 above)
Comment (markijbema) [reply src/services/ghost/classic-auto-complete/contextualSkip.ts:107]: sure, but it is a slight optimizement
